apps/omg/lib/omg/ethereum_event_listener/core.ex

```diff
@@ -70,89 +63,32 @@ defmodule OMG.EthereumEventListener.Core do
-          cached: %{request_max_size: request_max_size, events_upper_bound: old_upper_bound} = cached_data
-        } = state,
-        %SyncGuide{root_chain_height: root_chain_height, sync_height: sync_height}
-      ) do
-    # grab as much as allowed, but not higher than current root_chain_height and at least as much as needed to sync
-    # NOTE: both root_chain_height and sync_height are assumed to have any required finality margins applied by caller
-    next_upper_bound =
-      min(root_chain_height, old_upper_bound + request_max_size)
-      |> max(sync_height)
-
-    new_state = %__MODULE__{
-      state
-      | cached: %{cached_data | events_upper_bound: next_upper_bound}
-    }
-
-    {:get_events, {old_upper_bound + 1, next_upper_bound}, new_state}
-  end
-
-  @doc """
-  Stores the freshly fetched ethereum events into a memory-cache
-  """
-  @decorate span(service: :ethereum_event_listener, type: :backend, name: "add_new_events/2")
-  @spec add_new_events(t(), list(event)) :: t()
-  def add_new_events(
-        %__MODULE__{cached: %{data: data} = cached_data} = state,
-        new_events
-      ) do
-    %__MODULE__{state | cached: %{cached_data | data: data ++ new_events}}
-  end
-
-  @doc """
-  Pop some ethereum events stored in the memory-cache, up to a certain height
-  """
-  @decorate span(service: :ethereum_event_listener, type: :backend, name: "get_events/2")
-  @spec get_events(t(), non_neg_integer) :: {:ok, list(event), list(), non_neg_integer, t()}
-  def get_events(
-        %__MODULE__{synced_height_update_key: update_key, cached: %{data: data}} = state,
-        new_sync_height
-      ) do
-    {events, new_data} = Enum.split_while(data, fn %{eth_height: height} -> height <= new_sync_height end)
-
-    new_state =
-      state
-      |> Map.update!(:synced_height, &max(&1, new_sync_height))
-      |> Map.update!(:cached, &%{&1 | data: new_data})
-      |> struct!()
-
-    height_to_check_in = get_height_to_check_in(new_state)
-    db_update = [{:put, update_key, height_to_check_in}]
-    {:ok, events, db_update, height_to_check_in, new_state}
+  def get_events_range_for_download(state, sync_guide) do
+    case sync_guide.sync_height <= state.synced_height do
+      true ->
+        {:dont_fetch_events, state}
+
+      _ ->
+        # grab as much as allowed, but not higher than current root_chain_height and at least as much as needed to sync
+        # both root_chain_height and sync_height are assumed to have any required finality margins applied by caller
+        root_chain_height = sync_guide.root_chain_height
+        request_max_size = state.request_max_size
+        # root_chain_height is ethereums current height, we can't get pass that!
+        # so we find the min between root chain height and
+        # the current sync hight (state.synced_height) + what the max is (default is 1000)
+        next_upper_bound = max(min(root_chain_height, state.synced_height + request_max_size), sync_guide.sync_height)
```

Comment:
actually this formula isn't optimal, for example, if childchain (or watcher) would be restarted and falling behind ethereum, it would take *as much as it can* in terms of `request_max_size`!